src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,77 @@
 package mate.academy.lib;
 
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import mate.academy.service.impl.FileReaderServiceImpl;
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
     private static final Injector injector = new Injector();
+    private final Map<Class<?>, Object> instances = new HashMap<>();
 
     public static Injector getInjector() {
         return injector;
     }
 
-    public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+    public <T> T getInstance(Class<T> interfaceClass) {
+        if (instances.containsKey(interfaceClass)) {
+            return interfaceClass.cast(instances.get(interfaceClass));
+        }
+
+        for (Class<?> clazz : getAllClasses()) {
+            if (interfaceClass.isAssignableFrom(clazz)
+                    && clazz.isAnnotationPresent(Component.class)) {
+                try {
+                    T instance = (T) createInstance(clazz);
+                    instances.put(interfaceClass, instance);
+                    return instance;
+                } catch (Exception e) {
+                    throw new RuntimeException("Can't create object of the class", e);
+                }
+            }
+        }
+        throw new RuntimeException("No implementation found for " + interfaceClass.getName());
+    }
+
+    private <T> T createInstance(Class<T> clazz) {
+        if (!clazz.isAnnotationPresent(Component.class)) {
+            throw new RuntimeException("Class " + clazz.getName()
+                    + " is not annotated with @Component");
+        }
+
+        T instance;
+        try {
+            instance = clazz.getDeclaredConstructor().newInstance();
+        } catch (InstantiationException | IllegalAccessException
+                 | InvocationTargetException | NoSuchMethodException e) {
```

Comment:
please use their common ancestor

Reply:
![image](https://github.com/mate-academy/jv-dependency-injection/assets/71705897/4ed5f919-b3f3-4f11-a29a-f65bde1533cf)

But their common ancestor is 'Throwable': NoSuchMethodException is a subclass of 'Error', but InvocationTargetException is a subclass of 'Exception'